src/com/facebook/buck/rules/macros/QueryPathsMacroExpander.java

```diff
@@ -0,0 +1,131 @@
+package com.facebook.buck.rules.macros;
+
+import com.facebook.buck.model.BuildTarget;
+import com.facebook.buck.model.MacroException;
+import com.facebook.buck.query.QueryBuildTarget;
+import com.facebook.buck.query.QueryFileTarget;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.BuildRuleResolver;
+import com.facebook.buck.rules.CellPathResolver;
+import com.facebook.buck.rules.DefaultSourcePathResolver;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.rules.SourcePathRuleFinder;
+import com.facebook.buck.rules.TargetGraph;
+import com.facebook.buck.rules.query.Query;
+import com.facebook.buck.util.HumanReadableException;
+import com.facebook.buck.util.MoreCollectors;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.nio.file.Path;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class QueryPathsMacroExpander extends QueryMacroExpander<QueryPathsMacro> {
+
+  public QueryPathsMacroExpander(Optional<TargetGraph> targetGraph) {
+    super(targetGraph);
+  }
+
+  @Override
+  public Class<QueryPathsMacro> getInputClass() {
+    return QueryPathsMacro.class;
+  }
+
+  @Override
+  QueryPathsMacro fromQuery(Query query) {
+    return QueryPathsMacro.of(query);
+  }
+
+  @Override
+  boolean detectsTargetGraphOnlyDeps() {
+    return false;
+  }
+
+  @Override
+  public String expandFrom(
+      BuildTarget target,
+      CellPathResolver cellNames,
+      BuildRuleResolver resolver,
+      QueryPathsMacro input,
+      QueryResults precomputedWork) throws MacroException {
+    SourcePathResolver pathResolver =
+        DefaultSourcePathResolver.from(new SourcePathRuleFinder(resolver));
+
+    return precomputedWork
+        .results
+        .stream()
+        .map(
+            queryTarget -> {
+              // What we do depends on the input.
+              if (QueryBuildTarget.class.isAssignableFrom(queryTarget.getClass())) {
+                BuildRule rule = resolver.getRule(((QueryBuildTarget) queryTarget).getBuildTarget());
+                return Optional.ofNullable(rule.getSourcePathToOutput())
```

Comment:
What's the use case for the mixed-mode outputs/inputs here?

Reply:
You want people to be able to use a query string that returns either rules or files and have it work in the same way, which makes it easier to iteratively futz with the query string as you tweak a build to get it Just Right.

Ideally, there would be just a single `query` macro that works no matter what it is you're trying to do. That's not where we are :)